### PR TITLE
:hammer: Improve `mypy` Workflow and Fix Typing Errors

### PR DESCRIPTION
**Pull request by Jiaqi-Lv**
This pull request includes several bug fixes and improvements across utility functions, type checking, and testing in the codebase. The main focus is on improving error handling in image reading, refining type conversions, and enhancing test coverage for edge cases.

**Bug fixes and error handling improvements:**

* Improved `imread` in `utils/misc.py` to raise a `FileNotFoundError` when the image path does not exist and an `OSError` when OpenCV fails to read the image, ensuring clearer error reporting.
* Updated `cast_to_min_dtype` in `utils/misc.py` to use a tuple of (max_value, dtype) pairs for more robust and readable type selection.
* Fixed a bug in `dict_to_store_semantic_segmentor` by converting `layer_list` to a Python list before further processing, ensuring compatibility with downstream code. [[1]](diffhunk://#diff-dfa5a361c3c57e515728d23b53be87e0ee8da01ea04ae9cd6c39770fe3f1aaedR1480-R1487) [[2]](diffhunk://#diff-dfa5a361c3c57e515728d23b53be87e0ee8da01ea04ae9cd6c39770fe3f1aaedL1488-R1496)
* Refined coordinate type casting in `filter_coordinates` in `tools/patchextraction.py` to use `cast` for explicit type annotation.

**Testing improvements:**

* Added tests for `imread` to verify correct exceptions are raised for invalid paths and unreadable images, increasing test coverage for error scenarios.
* Added a test to ensure `compile_model` raises a `ValueError` when called with `None`, improving robustness of model compilation utilities.

**Type checking and code clarity:**

* Added a debug step to the MyPy type checking workflow and changed the import following mode to `silent` for better stability.
* Added missing imports and cleaned up imports in test files for clarity and correctness. [[1]](diffhunk://#diff-33c13e0b177bacd2f02e29bcb8aea5b49e7ce34901fd8f41fefb65defba1bd33R7) [[2]](diffhunk://#diff-33c13e0b177bacd2f02e29bcb8aea5b49e7ce34901fd8f41fefb65defba1bd33R46)
* Minor refactorings for clarity and explicitness, such as assigning the `attention` attribute in model utilities and improving variable naming in `tools/graph.py` and `utils/image.py`. [[1]](diffhunk://#diff-6eae2ccdde40d5d9c4749fb799b7777ceafbf1778f7b7f4b317af2972dde4225R328-R329) [[2]](diffhunk://#diff-a57a552474d66ef5961593e747fd21ba3fe25530b5a694dd5e8e467614bbc816L387-R391) [[3]](diffhunk://#diff-57f834801126808967286c68c43cb74bb25efaab0369210d0184e2576da66f3bL650-R655) [[4]](diffhunk://#diff-7fc02666e28590e37ff11a7ef254c9b2eb52eaf0f08017184d6e614df176ee80L6-R6)

**Model utility improvements:**

* Added an explicit check in `compile_model` to raise a `ValueError` if `model` is `None`, preventing silent failures.

These changes collectively improve code reliability, maintainability, and test coverage.

### CODE REVIEW

File: tiatoolbox/utils/misc.py [hunk 1]
```diff
@@ -1470,13 +1477,14 @@ def dict_to_store_semantic_segmentor(
     # Get the number of unique predictions
     layer_list = da.unique(preds).compute()
     layer_list = np.delete(layer_list, np.where(layer_list == ignore_index))
```
Comment (shaneahmed) [inline tiatoolbox/utils/misc.py]:
```suggestion
    layer_list = (np.delete(layer_list, np.where(layer_list == ignore_index))).tolist()
```

Can you not do as above? You do not need to create a new variable then. Also it does not sound good layer_list_list? Another way could be 

```python
layer_np = np.delete(layer_list, np.where(layer_list == ignore_index)))
layer_list = layer_np.tolist()
```

File: tiatoolbox/utils/image.py [hunk 2]
```diff
@@ -647,7 +647,12 @@ def sub_pixel_read(  # skipcq: PY-R1000  # noqa: C901, PLR0912, PLR0913, PLR0915
         ],
     )
     residuals = np.abs(int_read_bounds - read_bounds)
-    read_bounds = int_read_bounds
+    read_bounds = (
```
Comment (shaneahmed) [inline tiatoolbox/utils/image.py]:
```suggestion
    read_bounds: tuple = tuple(int_read_bounds)
```

File: tiatoolbox/tools/patchextraction.py [hunk 3]
```diff
@@ -432,7 +432,10 @@ def filter_coordinates(
             0,
             tissue_mask.shape[0],
         )
-        scaled_coords_list = list((scaled_coords).astype(np.int32))
+        scaled_coords_list = cast(
```
Comment (shaneahmed) [inline tiatoolbox/tools/patchextraction.py:435]:
```suggestion
        scaled_coords_list: list = (scaled_coords).astype(np.int32).to_list()
```

Why do you need to cast? Can you not do as above?

Comment (Jiaqi-Lv) [reply tiatoolbox/tools/patchextraction.py:435]: mypy still complains `scaled_coords_list: list = (scaled_coords).astype(np.int32).to_list()`, the only solution I found is to use `cast`.

File: tiatoolbox/utils/misc.py (hunk 1 above)
Comment (Jiaqi-Lv) [reply tiatoolbox/utils/misc.py]: I've updated it.

File: tiatoolbox/utils/image.py (hunk 2 above)
Comment (Jiaqi-Lv) [reply tiatoolbox/utils/image.py]: I've updated it.

**Review by shaneahmed — APPROVED** [review]
Thanks @Jiaqi-Lv
